File: src/tb_cascade/io.py
"""Phase 1 - Ingestion.

Reads the raw programmatic-surveillance CSV into a correctly typed
DataFrame, and writes an immutable, dated Parquet snapshot of whatever
input a given pipeline run used.

Dtype map below is built directly from
`Documentation/DataSet Description (English).md` and verified against the
real CSV header (62 columns, see `tests/test_io.py`).

Notes on dtype choices:
- All `0/1` flag columns use the pandas *nullable* ``"boolean"`` dtype
  rather than plain ``bool``, because the source export renders some
  flags as float-formatted text (``"1.0"``, ``"0.0"``) once a column
  contains any missing values, and a small number of flags are
  legitimately missing (e.g., not yet applicable to a given record).
- Numeric ID/count/code columns use nullable ``"Int64"`` for the same
  reason (e.g., ``FinalOutcome`` is blank for records without a recorded
  outcome).
- ``IndexCase`` is read as ``"string"`` rather than an integer type: values
  have meaningful leading zeros (e.g., ``"0107715"``) that a numeric dtype
  would silently destroy.
- ``Source`` is read as ``"category"`` per the implementation plan (a small,
  fixed set of site names).
- All `Date*` columns are parsed as ``datetime64[ns]`` via ``parse_dates``.
"""

from __future__ import annotations

from pathlib import Path

import pandas as pd

from tb_cascade.config import PROCESSED_DATA_DIR, RAW_DATA_PATH

# --- Column categorization (62 columns total) -------------------------------

# Nullable boolean ("0/1" flags), 39 columns.
BOOLEAN_COLUMNS: list[str] = [
    "Contact",
    "Homeless",
    "PLHIV",
    "Others",
    "TreatGroup_01",
    "TreatGroup_02",
    "TreatGroup_03",
    "Screening",
    "SuspectedTB",
    "DiaskintestPositive",
    "Screening_y",
    "NoTbAfter_y_xray",
    "NoTbAfter_y",
    "Screening_24",
    "NoTbAfter_24",
    "CompleteExaminationTB",
    "ConfirmedDiagnosisTB",
    "LTI",
    "NoTBNoLTI",
    "NoTBLTIunknown",
    "PrevTreatmentRec",
    "PrevTreatmentPresc",
    "PrevTreatmentStart",
    "RegBq",
    "RegMfx",
    "TreatmentCompleted",
    "TreatmentFinished",
    "TBdeveloped",
    "TreatmentStopedMed",
    "TreatmetnNotFinished",
    "TreatmentContinue",
    "OutcomeNotKnown",
    "Take50pc",
    "Take100pc",
    "SuppScreening",
    "Supp50pc",
    "Supp100pc",
    "Supp1yearGr23",
    "Supp1yearGr1",
]

# Nullable integer (IDs, codes, counts), 9 columns.
INT_COLUMNS: list[str] = [
    "Source_id",
    "Nomer",
    "Sex",
    "TargetGroup",
    "TreatGroup",
    "RelationWithSource",
    "DosesTaken",
    "SchemaDoses",
    "FinalOutcome",
]

# Fixed, small value set -> category, 1 column.
CATEGORY_COLUMNS: list[str] = [
    "Source",
]

# Free-text / ID with meaningful leading zeros, 1 column.
STRING_COLUMNS: list[str] = [
    "IndexCase",
]

# Parsed as datetime64[ns] via parse_dates, 12 columns.
DATE_COLUMNS: list[str] = [
    "BirthDate",
    "DateScreening",
    "DateScreening_y",
    "DateCompleteExaminationTB",
    "DatePrevTreatmentStart",
    "DateTreatmentScheme",
    "DateOutcome",
    "DateSuppScreening",
    "DateSupp50pc",
    "DateSupp100pc",
    "DateSupp1yearGr23",
    "DateSupp1yearGr1",
]

EXPECTED_COLUMN_COUNT = 62
assert (
    len(BOOLEAN_COLUMNS)
    + len(INT_COLUMNS)
    + len(CATEGORY_COLUMNS)
    + len(STRING_COLUMNS)
    + len(DATE_COLUMNS)
    == EXPECTED_COLUMN_COUNT
)

DTYPE_MAP: dict[str, str] = {
    **{c: "boolean" for c in BOOLEAN_COLUMNS},
    **{c: "Int64" for c in INT_COLUMNS},
    **{c: "category" for c in CATEGORY_COLUMNS},
    **{c: "string" for c in STRING_COLUMNS},
}


def load_raw(path: Path | str = RAW_DATA_PATH) -> pd.DataFrame:
    """Load the raw programmatic-surveillance CSV into a typed DataFrame.

    Applies the explicit dtype/parse_dates map above so downstream code
    never has to guess a column's type. Boolean and integer flag columns
    use pandas' nullable dtypes so missing values survive the read intact
    (rather than silently upcasting the whole column to float64).

    A small number of records contain a malformed/out-of-range date in a
    `Date*` field (e.g. a transposed-digit year such as "1201-09-20" or
    "201-08-01" - out of pandas' nanosecond-precision Timestamp range).
    When this happens, pandas' C parser cannot bulk-parse the whole
    column and silently leaves it as `object` dtype with raw strings
    instead of raising. To guarantee every `Date*` column comes out as
    `datetime64[ns]` per the implementation plan, any such column is
    re-parsed here with `errors="coerce"`, turning the unparseable value
    into `NaT`. This is a real upstream data-quality issue, not an
    ingestion bug - it is intentionally surfaced as a missing date rather
    than guessed at, so Phase 2 QC can flag it and a reviewer can correct
    it against the source record.
    """
    df = pd.read_csv(
        path,
        dtype=DTYPE_MAP,
        parse_dates=DATE_COLUMNS,
    )
    for col in DATE_COLUMNS:
        if not pd.api.types.is_datetime64_any_dtype(df[col]):
            df[col] = pd.to_datetime(df[col], errors="coerce")
    return df


def snapshot(df: pd.DataFrame, run_date: str) -> Path:
    """Write an immutable Parquet snapshot of ``df`` for a given run.

    The output path is ``Data/processed/raw_snapshot_<run_date>.parquet``.
    This lets every later pipeline stage and report cite the exact dataset
    version a run was built from, independent of the live raw CSV.

    Requires ``pyarrow`` (declared in ``pyproject.toml``).
    """
    PROCESSED_DATA_DIR.mkdir(parents=True, exist_ok=True)
    out_path = PROCESSED_DATA_DIR / f"raw_snapshot_{run_date}.parquet"
    df.to_parquet(out_path, index=False)
    return out_path

File: r/R/io.R
# Column layout of Data/raw/VladKovMur_dataset.csv, verified against the
# real file header and Documentation/DataSet Description (English).md.
# Order matters: it is also the expected-column-order regression guard.
.EXPECTED_COLUMNS <- c(
  "Source_id", "Source", "Nomer", "IndexCase", "BirthDate", "Sex",
  "TargetGroup", "Contact", "Homeless", "PLHIV", "Others", "TreatGroup",
  "TreatGroup_01", "TreatGroup_02", "TreatGroup_03", "RelationWithSource",
  "Screening", "DateScreening", "SuspectedTB", "DiaskintestPositive",
  "Screening_y", "DateScreening_y", "NoTbAfter_y_xray", "NoTbAfter_y",
  "Screening_24", "NoTbAfter_24", "CompleteExaminationTB",
  "DateCompleteExaminationTB", "ConfirmedDiagnosisTB", "LTI", "NoTBNoLTI",
  "NoTBLTIunknown", "PrevTreatmentRec", "PrevTreatmentPresc",
  "PrevTreatmentStart", "DatePrevTreatmentStart", "DateTreatmentScheme",
  "RegBq", "RegMfx", "TreatmentCompleted", "TreatmentFinished",
  "DateOutcome", "TBdeveloped", "TreatmentStopedMed",
  "TreatmetnNotFinished", "TreatmentContinue", "OutcomeNotKnown",
  "DosesTaken", "SchemaDoses", "Take50pc", "Take100pc",
  "DateSuppScreening", "SuppScreening", "DateSupp50pc", "Supp50pc",
  "DateSupp100pc", "Supp100pc", "DateSupp1yearGr23", "Supp1yearGr23",
  "DateSupp1yearGr1", "Supp1yearGr1", "FinalOutcome"
)

.CHARACTER_COLS <- c("Source", "IndexCase")

.DATE_COLS <- c(
  "BirthDate", "DateScreening", "DateScreening_y",
  "DateCompleteExaminationTB", "DatePrevTreatmentStart",
  "DateTreatmentScheme", "DateOutcome", "DateSuppScreening",
  "DateSupp50pc", "DateSupp100pc", "DateSupp1yearGr23", "DateSupp1yearGr1"
)

#' Column type specification for the raw dataset
#'
#' Every `0/1` flag and numeric code is read as `double` (the source export
#' writes some of them with a trailing `.0`, inconsistently across columns)
#' and then cast to integer in [load_raw()] -- not declared as integer here,
#' since `readr::col_integer()` rejects `"12.0"`-style literals outright.
#' Values are kept as raw coded integers (e.g. `TargetGroup` stays `1`-`4`);
#' recoding into labelled factors happens downstream in `derive.R`/`cascade.R`.
#' Date columns are read as character, not `col_date()`: a handful of rows
#' carry a `" 00:00:00"` timestamp suffix on an otherwise-plain date, which a
#' fixed-format date parser rejects outright; [load_raw()] parses them itself
#' with a more tolerant routine.
#'
#' @return A `readr::cols()` specification.
#' @keywords internal
raw_col_types <- function() {
  char_spec <- stats::setNames(
    rep(list(readr::col_character()), length(.CHARACTER_COLS)),
    .CHARACTER_COLS
  )
  date_spec <- stats::setNames(
    rep(list(readr::col_character()), length(.DATE_COLS)),
    .DATE_COLS
  )
  do.call(readr::cols, c(list(.default = readr::col_double()), char_spec, date_spec))
}

#' Parse a date column tolerant of a trailing time-of-day component
#'
#' Truncates to the first 10 characters (`YYYY-MM-DD`) before parsing, so
#' `"2020-05-14 00:00:00"` parses the same as `"2020-05-14"`. A string that is
#' still unparseable after truncation (e.g. a truncated year like
#' `"201-08-01"`) becomes `NA`, same as readr's own coercion behaviour, and is
#' reported by [load_raw()] rather than silently swallowed.
#'
#' Strings are required to match `YYYY-MM-DD` exactly (4-digit year) before
#' being handed to [as.Date()]: R's date parser otherwise accepts a
#' short/truncated year (e.g. `"201-08-01"` parses as year 201 rather than
#' failing), which would silently mask a real data-entry error as a valid
#' date far in the past.
#'
#' @param x Character vector.
#' @return A `Date` vector.
#' @keywords internal
.parse_date_col <- function(x) {
  cleaned <- substr(trimws(x), 1, 10)
  cleaned[!grepl("^[0-9]{4}-[0-9]{2}-[0-9]{2}$", cleaned)] <- NA
  as.Date(cleaned, format = "%Y-%m-%d")
}

#' Load the raw VladKovMur_dataset.csv export
#'
#' Reads with an explicit column-type map (see [raw_col_types()]), verifies
#' the column layout exactly matches the documented data dictionary, and
#' casts every flag/code/count column to integer. Row count is deliberately
#' *not* asserted here -- the dataset grows with each new export; pin row
#' count in a test against a specific snapshot instead, not in this
#' function.
#'
#' @param path Path to the CSV file. Defaults to [raw_data_path()].
#' @return A tibble with [N_COLS_EXPECTED] columns.
#' @export
load_raw <- function(path = raw_data_path()) {
  df <- readr::read_csv(
    path,
    col_types = raw_col_types(),
    na = c("", "NA"),
    progress = FALSE
  )

  actual_cols <- names(df)
  if (!identical(actual_cols, .EXPECTED_COLUMNS)) {
    missing <- setdiff(.EXPECTED_COLUMNS, actual_cols)
    extra <- setdiff(actual_cols, .EXPECTED_COLUMNS)
    stop(glue::glue(
      "Raw CSV column layout does not match the expected data dictionary.\n",
      "Missing: {paste(missing, collapse = ', ')}\n",
      "Unexpected: {paste(extra, collapse = ', ')}"
    ), call. = FALSE)
  }

  parsed_dates <- lapply(.DATE_COLS, function(col) .parse_date_col(df[[col]]))
  names(parsed_dates) <- .DATE_COLS
  malformed <- vapply(.DATE_COLS, function(col) {
    sum(!is.na(df[[col]]) & is.na(parsed_dates[[col]]))
  }, integer(1))
  malformed <- malformed[malformed > 0]
  if (length(malformed) > 0) {
    detail <- paste(sprintf("%s (%d)", names(malformed), malformed), collapse = ", ")
    warning(glue::glue(
      "load_raw(): {sum(malformed)} date value(s) were not parseable and ",
      "were coerced to NA: {detail}"
    ), call. = FALSE)
  }
  for (col in .DATE_COLS) {
    df[[col]] <- parsed_dates[[col]]
  }

  integer_cols <- setdiff(actual_cols, c(.CHARACTER_COLS, .DATE_COLS))
  df <- dplyr::mutate(df, dplyr::across(dplyr::all_of(integer_cols), as.integer))

  df
}

#' Write an immutable Parquet snapshot of a loaded raw dataset
#'
#' One snapshot per `run_date` so any later report can cite the exact input
#' it was built from.
#'
#' @param df Data frame to snapshot (typically the output of [load_raw()]).
#' @param run_date Date identifying this run; defaults to today.
#' @param dir Output directory; defaults to [processed_data_dir()].
#' @return The path the snapshot was written to, invisibly.
#' @export
snapshot <- function(df, run_date = Sys.Date(), dir = processed_data_dir()) {
  if (!dir.exists(dir)) {
    dir.create(dir, recursive = TRUE)
  }
  path <- file.path(dir, sprintf("raw_snapshot_%s.parquet", format(run_date, "%Y-%m-%d")))
  arrow::write_parquet(df, path)
  invisible(path)
}
